addons/l10n_pe_website_sale/controllers/main.py

```diff
@@ -0,0 +1,86 @@
+        company_country_id = request.website.company_id.country_id.id
+        pe_country_id = request.env.ref("base.pe").id
+        values_for_localization = ["city_id", "l10n_pe_district", "l10n_latam_identification_type_id", "vat"]
+        if country_id == pe_country_id and company_country_id == pe_country_id:
+            res += values_for_localization
+            res.remove("city")
+        return res
+
+    def _get_mandatory_fields_shipping(self, country_id=False):
+        """Extend mandatory fields to add city_id and district fields when the selected country is Peru"""
+        res = super()._get_mandatory_fields_shipping(country_id)
+        company_country_id = request.website.company_id.country_id.id
+        pe_country_id = request.env.ref("base.pe").id
+        values_for_localization = ["city_id", "l10n_pe_district"]
+        if country_id == pe_country_id and company_country_id == pe_country_id:
+            res += values_for_localization
+            res.remove("city")
+        return res
+
+    def _get_country_related_render_values(self, kw, render_values):
+        res = super()._get_country_related_render_values(kw, render_values)
+
+        if request.website.sudo().company_id.country_id.code == "PE":
+            values = render_values["checkout"]
+            state = "state_id" in values \
+                    and values["state_id"] != "" \
+                    and request.env["res.country.state"].browse(int(values["state_id"]))
+            city = "city_id" in values \
+                    and values["city_id"] != "" \
+                    and request.env["res.city"].browse(int(values["city_id"]))
+            to_include = {
+                "identification": kw.get("l10n_latam_identification_type_id"),
+                "identification_types": request.env["l10n_latam.identification.type"].sudo().search(
+                    ["|", ("country_id", "=", False), ("country_id.code", "=", "PE")]
+                ),
+            }
+            if state:
+                to_include["state"] = state
+                to_include["state_cities"] = request.env["res.city"].sudo().search([("state_id", "=", state.id)])
+            if city:
+                to_include["city"] = city
+                to_include["city_districts"] = request.env["l10n_pe.res.city.district"].sudo().search([("city_id", "=", city.id)])
+            res.update(to_include)
+        return res
+
+    def _get_vat_validation_fields(self, data):
+        res = super()._get_vat_validation_fields(data)
+        if request.website.sudo().company_id.account_fiscal_country_id.code == "PE":
+            res.update({
+                "l10n_latam_identification_type_id":
+                    int(data["l10n_latam_identification_type_id"])
+                    if data.get("l10n_latam_identification_type_id") else False,
+                "name": data.get("name", False),
+            })
+        return res
+
+    @http.route(
```

Comment:
There should not be a need for this

Reply:
Hello @jco-odoo 

Do you refer to the method? And move this section to `_onChangeState` function?

Reply:
@luistorresm I believe it should be possible to call the search immediately from the javascript instead of having to define a separate controller or method.

Reply:
Hello @jco-odoo 

I tested some options, but I did not get an option to avoid this, because is not a component.

You can see that the base module use the same way: https://github.com/odoo/odoo/blob/17.0/addons/website_sale/static/src/js/website_sale.js#L206

Reply:
@jco-odoo we were looking at the documentation and the code looking for a way to directly call a method as we did in v16 and prior versions, It seems that this is no longer possible in v17.
as you can see the RPC to a public controller  is made the same way to get the states of a country in the checkout here: https://github.com/odoo/odoo/blob/17.0/addons/website_sale/static/src/js/website_sale.js#L240

To directly call a method like search or search_read from the website you need to have available the ORM hook, which is only available in OWL components, this is not the case of the JS module ([WebsiteSale](https://github.com/odoo/odoo/blob/17.0/addons/website_sale/static/src/js/website_sale.js#L15))

please let me know if you know a possible way to do it without using OWL components.

cc @luistorresm

Reply:
@jco-odoo Surprising, it seems website_sale has not yet been fully migrated to owl - most of our code seems jquery based - old school. I guess @Feyensv would be best to comment on this

Reply:
AFAIK, the ability to call search/search_read/other orm methods is limited to logged in users, whereas e-commerce flows are available to public users.
That's why dedicated routes are needed, independently of owl conversion or whatever.